content/dev/blog/tidyeval.Rmd

```diff
@@ -0,0 +1,232 @@
+`autoplot()` was the ideal approach for us to encapsulate the logic of how to "just do the right thing." You can do the analysis you want, and it will make smart choices about how to display it with no additional input--all of which you can control or override with additional `ggplot2` layers, if you want.
+
+This lets us do [insert examples here, with pictures]
+
+Making it look easy, though, can be hard work. Our `autoplot` methods inspect the input objects to understand their dimensionality and data types and choose an appropriate visualization. Figuring out and passing the right arguments to the right places can be messy, so make these functions work, we took advantage of "[tidyeval](https://dplyr.tidyverse.org/articles/programming.html)", a new framework that systematizes non-standard evaluation in R.
+
+To illustrate this pattern, this blog post goes through a simplified example using the "diamonds" example dataset and pure `dplyr` methods. We'll create a plotting function that adjusts to the number of grouping variables in a tibble. This lets you pipe data from a dplyr pipeline into a single function and get a meaningful, appropriate plot, like this:
+
+```{r}
+library(ggplot2)
+library(dplyr)
+diamonds %>%
+    group_by(cut, clarity) %>%
+    tally() %>%
+    autoplot()
+
+diamonds %>%
+    group_by(clarity, cut, color) %>%
+    summarize(mean_price = mean(price)) %>%
+    autoplot()
+```
+
+The actual code for our Crunch `autoplot` methods is [here](https://github.com/Crunch-io/crplyr/blob/ggplot-methods/R/plotting.R).
+
+## Non-Standard Evaluation: Great Power, Great Pain
+
+A great feature the R language is that it lets you access and manipulate the environment in which a function is called. This non-standard evaluation (NSE) gives package authors a flexible and powerful way to build programming interfaces.
+
+At Crunch, we use NSE a lot. One example is in reporting better error messages. If you send the wrong data to the API, you might get back a response like `400: Payload is malformed`. That's not very helpful even for users who know our API well. We use NSE to inspect the user's calling environment, figure out which variables or data structures are causing the problem, and suggest a fix.
+
+Despite working this framework all the time, I'm pretty sure that I've never once gotten it right on the first try. The core reason is that whenever you are capturing and expression to evaluate later you need to also keep track of which environment you should evaluate that expression in. This makes it really difficult to pass unevaluated expressions between functions and have the evaluation occur without error. For instance, take this code which replaces a `missing` function argument with a logical value:
```

Comment:
I wonder if there's a shorter example of this (perhaps using the error handling code mentioned above?), or whether we need to provide an example at all. Is it necessary for us to prove that NSE is hard, or can we just say that we always f it up the first time we try because scoping and environments are hard.